src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs

```diff
@@ -427,6 +429,94 @@ public override bool ValidateChildren
             }
         }
 
+        /// <inheritdoc />
+        public override bool? HasValidators
+        {
+            get
+            {
+                if (!_hasValidators.HasValue)
+                {
+                    var visited = new HashSet<DefaultModelMetadata>();
+
+                    if (TryCalculateHasValidators(visited, this, out var hasValidators))
+                    {
+                        _hasValidators = hasValidators;
+                    }
+                    else
+                    {
+                        // Assume validation is required by default.
+                        _hasValidators = true;
+                    }
+                }
+
+                return _hasValidators.Value;
+            }
+        }
+
+        internal static bool TryCalculateHasValidators(HashSet<DefaultModelMetadata> visited, ModelMetadata metadata, out bool hasValidators)
```

Comment:
Discussed offline. I think this would be much less complicated seeming if it were not a `Try...` you can just return false if you detect a cycle.